azkaban-common/src/main/java/azkaban/executor/container/ContainerCleanupManager.java

```diff
@@ -0,0 +1,132 @@
+  private final ScheduledExecutorService cleanupService;
+  private final ExecutorLoader executorLoader;
+  private final ContainerizedImpl containerizedImpl;
+
+  @Inject
+  public ContainerCleanupManager(final Props azkProps, final ExecutorLoader executorLoader,
+      final ContainerizedImpl containerizedImpl) {
+    this.cleanupIntervalMin = azkProps
+        .getLong(
+            ContainerizedExecutionManagerProperties.CONTAINERIZED_STALE_EXECUTION_CLEANUP_INTERVAL_MIN,
+            DEFAULT_STALE_CONTAINER_CLEANUP_INTERVAL.toMinutes());
+    this.staleContainerAgeDays = azkProps
+        .getLong(
+            ContainerizedExecutionManagerProperties.CONTAINERIZED_STALE_EXECUTION_AGE_DAYS,
+            DEFAULT_STALE_CONTAINER_AGE.toDays());
+    this.cleanupService = Executors.newSingleThreadScheduledExecutor(
+        new ThreadFactoryBuilder().setNameFormat("azk-container-cleanup").build());
+    this.executorLoader = executorLoader;
+    this.containerizedImpl = containerizedImpl;
+  }
+
+  /**
+   * Execute container-provider specific APIs for all 'stale' containers. A container is considered
+   * 'stale' if it was launched CONTAINERIZED_STALE_EXECUTION_AGE_DAYS ago and the corresponding
+   * execution is not yet in a final state.
+   * <p>
+   * It's important that this method does not throw exceptions as that will interrupt the scheduling
+   * of {@code cleanupService}.
+   */
+  public void terminateStaleContainers() {
+    try {
+      final List<ExecutableFlow> staleFlows = this.executorLoader.fetchStaleFlows(Duration.ofDays(
+          this.staleContainerAgeDays));
+      for (final ExecutableFlow flow : staleFlows) {
+        deleteContainerQuietly(flow.getExecutionId());
+      }
+    } catch (final Exception e) {
+      logger.error("Unexpected exception during container cleanup." + e);
+    }
+  }
+
+  // Deletes the container specified by executionId while logging and consuming any exceptions.
+  // Note that while this method is not async it's still expected to return 'quickly'. This is true
+  // for Kubernetes as it's declarative API will only submit the request for deleting container
+  // resources. In future we can consider making this async to eliminate any chance of the cleanup
+  // thread getting blocked.
+  private void deleteContainerQuietly(final int executionId) {
+    try {
+      this.containerizedImpl.deleteContainer(executionId);
+    } catch (final ExecutorManagerException eme) {
+      logger.warn("ExecutorManagerException while deleting container.", eme);
+    } catch (final RuntimeException re) {
+      logger.error("Unexpected RuntimeException while deleting container.", re);
+    }
+  }
+
+  /**
+   * Start periodic deletions of per-container resources for stale executions.
+   */
+  @SuppressWarnings("FutureReturnValueIgnored")
+  public void start() {
+    logger.info("Start container cleanup service");
+    this.cleanupService.scheduleAtFixedRate(this::terminateStaleContainers, 0L,
```

Comment:
Why don't you use scheduleWithFixedDelay instead of scheduleAtFixedRate?

Reply:
`scheduleAtFixedRate` will provide better consistency for the periodicity of the cleanup check. With  `scheduleWithFixedDelay` any unexpected delays in the executions of the cleanup checks and/or the `deleteContainer` procedure will result in an avoidable delay in the next execution.

Reply:
Until the previous run has finished, if you will start another run then it will bring similar execution if it is not yet processed in previous run and status is not yet updated in db. How will you handle that?

Reply:
As discussed, `scheduleAtFixedRate` guarantees no concurrent executions will take place, so we should run into this.